lib/jxl/enc_fast_lossless.cc

```diff
@@ -3657,80 +3748,188 @@ JxlFastLosslessFrameState* LLEnc(const unsigned char* rgba, size_t width,
-
+  size_t num_dc_groups = num_dc_groups_x * num_dc_groups_y;
+  size_t num_ac_groups = num_groups_x * num_groups_y;
+  size_t num_groups = onegroup ? 1 : (2 + num_dc_groups + num_ac_groups);
+  frame_state->input = input;
   frame_state->width = width;
   frame_state->height = height;
+  frame_state->num_groups_x = num_groups_x;
+  frame_state->num_groups_y = num_groups_y;
+  frame_state->num_dc_groups_x = num_dc_groups_x;
+  frame_state->num_dc_groups_y = num_dc_groups_y;
   frame_state->nb_chans = nb_chans;
   frame_state->bitdepth = bitdepth.bitdepth;
+  frame_state->big_endian = big_endian;
+  frame_state->effort = effort;
+  frame_state->collided = collided;
+  frame_state->lookup = lookup;
 
   frame_state->group_data = std::vector<std::array<BitWriter, 4>>(num_groups);
+  frame_state->group_sizes.resize(num_groups);
   if (collided) {
-    PrepareDCGlobal(onegroup, width, height, nb_chans, hcode,
+    PrepareDCGlobal(onegroup, width, height, nb_chans, frame_state->hcode,
                     &frame_state->group_data[0][0]);
   } else {
-    PrepareDCGlobalPalette(onegroup, width, height, nb_chans, hcode, palette,
-                           pcolors, &frame_state->group_data[0][0]);
+    PrepareDCGlobalPalette(onegroup, width, height, nb_chans,
+                           frame_state->hcode, palette, pcolors,
+                           &frame_state->group_data[0][0]);
+  }
+  frame_state->group_sizes[0] = SectionSize(frame_state->group_data[0]);
+  if (!onegroup) {
+    ComputeAcGroupDataOffset(frame_state->group_sizes[0], num_dc_groups,
+                             num_ac_groups, frame_state->min_dc_global_size,
+                             frame_state->ac_group_data_offset);
   }
 
-  auto run_one = [&](size_t g) {
-    size_t xg = g % num_groups_x;
-    size_t yg = g / num_groups_x;
-    size_t group_id =
-        onegroup ? 0 : (2 + num_dc_groups_x * num_dc_groups_y + g);
-    size_t xs = std::min<size_t>(width - xg * 256, 256);
-    size_t ys = std::min<size_t>(height - yg * 256, 256);
-    size_t x0 = xg * 256;
-    size_t y0 = yg * 256;
-    auto& gd = frame_state->group_data[group_id];
-    if (collided) {
-      WriteACSection(rgba, x0, y0, xs, ys, stride, onegroup, bitdepth, nb_chans,
-                     big_endian, hcode, gd);
+  return frame_state;
+}
 
-    } else {
-      WriteACSectionPalette(rgba, x0, y0, xs, ys, stride, onegroup, hcode,
-                            lookup.data(), nb_chans, gd[0]);
+template <typename BitDepth>
+void LLProcess(JxlFastLosslessFrameState* frame_state, bool is_last,
+               BitDepth bitdepth, void* runner_opaque,
+               FJxlParallelRunner runner,
+               JxlEncoderOutputProcessorWrapper* output_processor) {
+  if (frame_state->process_done) {
+    JxlFastLosslessPrepareHeader(frame_state, /*add_image_header=*/0, is_last);
+    if (output_processor) {
+      JxlFastLosslessOutputFrame(frame_state, output_processor);
```

Comment:
In principle, one could start writing output even before all groups are done (i.e. one could flush out a prefix of the groups).

Reply:
This code-path is for the non-streaming input case where all the frames were processed already and we just generate the output.

Reply:
Ah, I must have missed that then :)